src/System.Net.Sockets.Legacy/tests/FunctionalTests/SendReceive.cs

```diff
@@ -0,0 +1,555 @@
+            int receivedDatagrams = -1;
+
+            Action<int, EndPoint> receiveHandler = null;
+            receiveHandler = (received, remote) =>
+            {
+                if (receivedDatagrams != -1)
+                {
+                    Assert.Equal(DatagramSize, received);
+                    Assert.Equal(rightEndpoint, remote);
+                    receivedChecksums[(int)receiveBuffer[0]] = Fletcher32.Checksum(receiveBuffer, 0, received);
+
+                    receiverAck.Set();
+                    senderAck.Wait();
+                    senderAck.Reset();
+
+                    receivedDatagrams++;
+                    if (receivedDatagrams == DatagramsToSend)
+                    {
+                        left.Dispose();
+                        receiverFinished.SetResult(true);
+                        return;
+                    }
+                }
+                else
+                {
+                    receivedDatagrams = 0;
+                }
+
+                left.ReceiveFromAPM(receiveBuffer, 0, receiveBuffer.Length, SocketFlags.None, receiveRemote, receiveHandler);
+            };
+            receiveHandler(0, null);
+
+            var random = new Random();
+            var sentChecksums = new uint[DatagramsToSend];
+            var sendBuffer = new byte[DatagramSize];
+            int sentDatagrams = -1;
+
+            Action<int> sendHandler = null;
+            sendHandler = sent =>
+            {
+                if (sentDatagrams != -1)
+                {
+                    receiverAck.Wait();
+                    receiverAck.Reset();
+                    senderAck.Set();
+
+                    Assert.Equal(DatagramSize, sent);
+                    sentChecksums[sentDatagrams] = Fletcher32.Checksum(sendBuffer, 0, sent);
+
+                    sentDatagrams++;
+                    if (sentDatagrams == DatagramsToSend)
+                    {
+                        right.Dispose();
+                        return;
+                    }
+                }
+                else
+                {
+                    sentDatagrams = 0;
+                }
+
+                random.NextBytes(sendBuffer);
+                sendBuffer[0] = (byte)sentDatagrams;
+                right.SendToAPM(sendBuffer, 0, sendBuffer.Length, SocketFlags.None, leftEndpoint, sendHandler);
+            };
+            sendHandler(0);
+
+            receiverFinished.Task.Wait();
+            for (int i = 0; i < DatagramsToSend; i++)
+            {
+                Assert.Equal(sentChecksums[i], receivedChecksums[i]);
+            }
+        }
+
+        private static void SendRecv_Stream_TCP(IPAddress listenAt, bool useMultipleBuffers)
+        {
+            const int BytesToSend = 123456;
+
+            var server = new Socket(listenAt.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+            server.BindToAnonymousPort(listenAt);
+
+            server.Listen(128);
```

Comment:
Nit: Usually better to name these consts to make the tests easier to tweak if we find that the CI system is too slow to handle it.